test_util/cluster.py

```diff
@@ -278,6 +358,118 @@ def install_dcos(
         installer.postflight()
 
 
+def upgrade_dcos(cluster, installer_url, add_config_path=None):
+
+    def master_upgrade_order(cluster):
+        """Return a list of masters with the ZooKeeper leader last."""
+        return [master_zk_mode[0] for master_zk_mode in sorted(
+            cluster.master_zk_modes(),
+            # False < True
+            key=lambda host_zk_mode: host_zk_mode[1] == 'leader',
+        )]
+
+    def upgrade_host(tunnel, role, bootstrap_url):
+        # Download the install script for the new DC/OS.
+        tunnel.remote_cmd([
+            'curl',
+            '--silent',
+            '--verbose',
+            '--show-error',
+            '--fail',
+            '--location',
+            '--keepalive-time', '2',
+            '--retry', '20',
+            '--speed-limit', '100000',
+            '--speed-time', '60',
+            '--remote-name',
+            bootstrap_url + '/dcos_install.sh',
+        ])
+
+        # Remove the old DC/OS.
+        tunnel.remote_cmd(['sudo', '-i', '/opt/mesosphere/bin/pkgpanda', 'uninstall'])
+        tunnel.remote_cmd(['sudo', 'rm', '-rf', '/opt/mesosphere', '/etc/mesosphere'])
+
+        # Install the new DC/OS.
+        tunnel.remote_cmd(['sudo', 'bash', 'dcos_install.sh', '-d', role])
```

Comment:
IIRC it'd complain about a lingering port that hadn't been cleaned up, maybe Exhibitor.